relayer/txm/txstore.go

```diff
@@ -44,22 +70,127 @@ func (s *TxStore) AddUnconfirmed(hash string, expirationMs int64, tx *TronTx) er
 	return nil
 }
 
-func (s *TxStore) Confirm(hash string) error {
+func (s *TxStore) OnBroadcasted(id string) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	tx, exists := s.unconfirmedTxs[id]
+	if !exists {
+		return fmt.Errorf("no such unconfirmed id: %s", id)
+	}
+	tx.Tx.State = Broadcasted
+
+	return nil
+}
+
+func (s *TxStore) OnConfirmed(id string) error {
 	s.lock.Lock()
 	defer s.lock.Unlock()
 
-	if _, exists := s.unconfirmedTxes[hash]; !exists {
-		return fmt.Errorf("no such unconfirmed hash: %s", hash)
+	tx, exists := s.unconfirmedTxs[id]
+	if !exists {
+		return fmt.Errorf("no such unconfirmed id: %s", id)
 	}
-	delete(s.unconfirmedTxes, hash)
+	delete(s.unconfirmedTxs, id)
+
+	tx.Tx.State = Confirmed
+	s.confirmedTxs[id] = tx
+
 	return nil
 }
 
-func (s *TxStore) GetUnconfirmed() []*UnconfirmedTx {
+func (s *TxStore) OnErrored(id string) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if pt, exists := s.unconfirmedTxs[id]; exists {
+		delete(s.hashToId, pt.Hash)
+		pt.Tx.State = Errored
+		return nil
+	}
+
+	// check if the tx is confirmed for sanity
+	if pt, exists := s.confirmedTxs[id]; exists {
+		delete(s.confirmedTxs, id)
+		delete(s.hashToId, pt.Hash)
+		pt.Tx.State = Errored
+		s.unconfirmedTxs[id] = pt
+		return nil
+	}
+
+	return fmt.Errorf("no such unconfirmed or confirmed id: %s", id)
+}
+
+func (s *TxStore) OnFatalError(id string) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	var pt *PendingTx
+	if _, unconfirmedExists := s.unconfirmedTxs[id]; !unconfirmedExists {
+		if _, confirmedExists := s.confirmedTxs[id]; !confirmedExists {
+			return fmt.Errorf("no such unconfirmed or confirmed id: %s", id)
+		} else {
+			pt = s.confirmedTxs[id]
+			delete(s.confirmedTxs, id)
+		}
+	} else {
+		pt = s.unconfirmedTxs[id]
+		delete(s.unconfirmedTxs, id)
+	}
```

Comment:
nit: couldn't we reformat this as 
```suggestion
	if pt, unconfirmedExists := s.unconfirmedTxs[id]; unconfirmedExists {
		delete(s.unconfirmedTxs, id)
	} else if pt, confirmedExists := s.confirmedTxs[id]; confirmedExists {
	        delete(s.confirmedTxs, id)
	} else {
	        return fmt.Errorf("no such unconfirmed or confirmed id: %s", id)
	}
```